File: app/src/main/java/com/example/islandlyrics/MainActivity.java
package com.example.islandlyrics;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.google.android.material.card.MaterialCardView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.graphics.Insets;

public class MainActivity extends BaseActivity {
    private static final String TAG = "IslandLyrics";
    private static final String PREFS_NAME = "IslandLyricsPrefs";
    private static final String PREF_WHITELIST = "whitelist_packages";

    // UI Elements
    private MaterialCardView mCardStatus;
    private ImageView mIvStatusIcon;
    private TextView mTvStatusText;
    private TextView mTvAppVersion;
    
    private TextView mTvSong;
    private TextView mTvArtist;
    private TextView mTvLyric;
    
    // Default Whitelist
    private static final String[] DEFAULT_WHITELIST = {
        "com.netease.cloudmusic",
        "com.tencent.qqmusic",
        "com.kugou.android",
        "com.spotify.music",
        "com.apple.android.music",
        "com.google.android.youtube",
        "com.google.android.apps.youtube.music"
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Handle Edge-to-Edge / Status Bar Insets
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        initializeDefaultWhitelist();
        bindViews();
        setupClickListeners();
        setupObservers();
        
        // NO Auto-Permission Checks on Launch (Privacy First)
        // Permissions are now handled in SettingsActivity with Disclaimer.
        
        updateVersionInfo();
    }
    
    // onResume is overridden below for Progress Logic
    
    private void initializeDefaultWhitelist() {
        SharedPreferences prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        if (!prefs.contains(PREF_WHITELIST)) {
            Set<String> set = new HashSet<>(Arrays.asList(DEFAULT_WHITELIST));
            prefs.edit().putStringSet(PREF_WHITELIST, set).apply();
        }
    }

    private void bindViews() {
        mCardStatus = findViewById(R.id.cv_status);
        mIvStatusIcon = findViewById(R.id.iv_status_icon);
        mTvStatusText = findViewById(R.id.tv_status_text);
        
        mTvAppVersion = findViewById(R.id.tv_app_version);
        
        mTvSong = findViewById(R.id.tv_song);
        mTvArtist = findViewById(R.id.tv_artist);
        mTvLyric = findViewById(R.id.tv_lyric);
        
        mPbDebug = findViewById(R.id.pb_debug_progress);
        mTvDebugTime = findViewById(R.id.tv_debug_time);
        mTvDebugInfo = findViewById(R.id.tv_debug_song_info);
    }

    private void setupClickListeners() {
        // Status Card -> Settings (Diagnostics moved to Settings)
        // mCardStatus.setOnClickListener(v -> startActivity(new Intent(this, SettingsActivity.class))); // Removed per new design, purely status now? Or keep?
        // User requested Settings Card at bottom.
        
        View cvSettings = findViewById(R.id.cv_settings);
        if (cvSettings != null) {
            cvSettings.setOnClickListener(v -> startActivity(new Intent(this, SettingsActivity.class)));
        }
        
        findViewById(R.id.btn_copy_lyric).setOnClickListener(v -> {
            String text = mTvLyric.getText().toString();
            copyToClipboard("Lyric", text);
        });

        findViewById(R.id.btn_copy_all).setOnClickListener(v -> {
            String song = mTvSong.getText().toString();
            String artist = mTvArtist.getText().toString();
            String lyric = mTvLyric.getText().toString();
            String full = song + " - " + artist + "\n" + lyric;
            copyToClipboard("Song Info", full);
        });
    }

    private void copyToClipboard(String label, String text) {
        if (text != null && !text.isEmpty() && !text.equals("-")) {
            android.content.ClipboardManager clipboard = (android.content.ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
            android.content.ClipData clip = android.content.ClipData.newPlainText(label, text);
            clipboard.setPrimaryClip(clip);
            Toast.makeText(this, label + " copied!", Toast.LENGTH_SHORT).show();
        }
    }

    private void updateStatusCardState() {
        // Check Permission
        boolean listenerGranted = isNotificationListenerEnabled();
        
        if (!listenerGranted) {
            // State: Missing Permission
            setCardState(false, "Permission Required", R.color.status_inactive); // Red/Orange
            // Optional: Blur playback card or show warning
        } else {
             // Check Playing Status
             Boolean isPlaying = LyricRepository.getInstance().getIsPlaying().getValue();
             if (Boolean.TRUE.equals(isPlaying)) {
                 String source = LyricRepository.getInstance().getSourceAppName().getValue();
                 setCardState(true, "Active: " + (source != null ? source : "Music"), R.color.status_active);
             } else {
                 setCardState(true, "Service Ready (Idle)", R.color.status_active); // Or distinct color?
             }
        }
    }
    
    private void setCardState(boolean isActive, String text, int colorResId) {
        mCardStatus.setCardBackgroundColor(ContextCompat.getColor(this, colorResId));
        mTvStatusText.setText(text);
        
        if (isActive) {
             mIvStatusIcon.setImageResource(R.drawable.ic_check_circle);
        } else {
             mIvStatusIcon.setImageResource(R.drawable.ic_cancel);
        }
    }
    
    private boolean isNotificationListenerEnabled() {
        return android.provider.Settings.Secure.getString(getContentResolver(),
                "enabled_notification_listeners").contains(getPackageName());
    }

    private void updateVersionInfo() {
        try {
            String version = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
            String type = (getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0 ? "Debug" : "Release";
            mTvAppVersion.setText(getString(R.string.app_version_fmt, version, type));
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
    }

    // Debug Progress UI
    private android.widget.ProgressBar mPbDebug;
    private TextView mTvDebugTime;
    private TextView mTvDebugInfo;
    private final android.os.Handler mHandler = new android.os.Handler();
    private final Runnable mProgressRunnable = new Runnable() {
        @Override
        public void run() {
            updateDebugProgress();
            mHandler.postDelayed(this, 1000);
        }
    };

    @Override
    protected void onResume() {
        super.onResume();
        updateStatusCardState();
        mHandler.post(mProgressRunnable);
    }

    @Override
    protected void onPause() {
        super.onPause();
        mHandler.removeCallbacks(mProgressRunnable);
    }

    private void updateDebugProgress() {
        if (!isNotificationListenerEnabled()) {
             mTvDebugInfo.setText("Permission Missing");
             return;
        }

        try {
            android.media.session.MediaSessionManager mm = (android.media.session.MediaSessionManager) getSystemService(MEDIA_SESSION_SERVICE);
            android.content.ComponentName component = new android.content.ComponentName(this, MediaMonitorService.class); // Use ANY component that has permission really
            // Note: getActiveSessions needs the ComponentName of the notification listener usually
            List<android.media.session.MediaController> controllers = mm.getActiveSessions(component);
            
            android.media.session.MediaController activeController = null;
            for (android.media.session.MediaController c : controllers) {
                if (c.getPlaybackState() != null && c.getPlaybackState().getState() == android.media.session.PlaybackState.STATE_PLAYING) {
                    activeController = c;
                    break;
                }
            }

            if (activeController != null) {
                android.media.session.PlaybackState state = activeController.getPlaybackState();
                android.media.MediaMetadata meta = activeController.getMetadata();
                
                long duration = 0;
                if (meta != null) {
                    duration = meta.getLong(android.media.MediaMetadata.METADATA_KEY_DURATION);
                }

                if (state != null) {
                    long lastPosition = state.getPosition();
                    long lastUpdateTime = state.getLastPositionUpdateTime();
                    float speed = state.getPlaybackSpeed();
                    
                    long currentPos = lastPosition + (long) ((android.os.SystemClock.elapsedRealtime() - lastUpdateTime) * speed);
                    
                    // Clamp
                    if (duration > 0 && currentPos > duration) currentPos = duration;
                    if (currentPos < 0) currentPos = 0; // Seek edge case

                    // Update UI
                    mTvDebugInfo.setText("Playing: " + activeController.getPackageName());
                    if (duration > 0) {
                        int progress = (int) (1000 * currentPos / duration);
                        mPbDebug.setProgress(progress);
                        mTvDebugTime.setText(formatTime(currentPos) + " / " + formatTime(duration));
                    } else {
                        mPbDebug.setProgress(0);
                        mTvDebugTime.setText(formatTime(currentPos) + " / --:--");
                    }
                }
            } else {
                mTvDebugInfo.setText("No Active Media");
            }

        } catch (SecurityException e) {
            mTvDebugInfo.setText("Security Exception (Check Permission)");
        } catch (Exception e) {
            mTvDebugInfo.setText("Error: " + e.getMessage());
        }
    }

    private String formatTime(long ms) {
        long seconds = ms / 1000;
        long minutes = seconds / 60;
        seconds = seconds % 60;
        return String.format("%02d:%02d", minutes, seconds);
    }
    
    private void setupObservers() {
        LyricRepository repo = LyricRepository.getInstance();

        // Playback Status Observer
        repo.getIsPlaying().observe(this, isPlaying -> {
            updateStatusCardState();
        });

        // Lyric Observer
        repo.getCurrentLyric().observe(this, lyric -> {
            if (lyric != null) {
                mTvLyric.setText(lyric);
            } else {
                mTvLyric.setText("-");
            }
        });

        // Metadata Observer
        repo.getLiveMetadata().observe(this, mediaInfo -> {
             if (mediaInfo == null) return;
             // Ensure defaults if null
             mTvSong.setText(mediaInfo.title != null ? mediaInfo.title : "-");
             mTvArtist.setText(mediaInfo.artist != null ? mediaInfo.artist : "-");
        });
    }

}

File: app/src/main/java/com/example/islandlyrics/MainActivity.kt
package com.example.islandlyrics

import android.content.BroadcastReceiver
import java.util.Arrays
import android.content.ClipboardManager
import android.content.ClipData
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView

class MainActivity : BaseActivity() {

    // UI Elements
    private lateinit var cardStatus: MaterialCardView
    private lateinit var ivStatusIcon: ImageView
    private lateinit var tvStatusText: TextView
    private lateinit var tvAppVersion: TextView

    private lateinit var tvSong: TextView
    private lateinit var tvArtist: TextView
    private lateinit var tvLyric: TextView

    // Dashboard UI
    private lateinit var tvApiPermission: TextView
    private lateinit var tvNotifCapability: TextView
    private lateinit var tvNotifFlag: TextView
    private lateinit var btnOpenPromotedSettings: MaterialButton
    private lateinit var pbProgress: com.google.android.material.progressindicator.LinearProgressIndicator

    private val handler = Handler(Looper.getMainLooper())


    private val diagReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if ("com.example.islandlyrics.DIAG_UPDATE" == intent.action) {
                // ... logic for raw logs if needed
            } else if ("com.example.islandlyrics.STATUS_UPDATE" == intent.action) {

                // General Service Status
                val status = intent.getStringExtra("status")
                val tvStatus = findViewById<TextView>(R.id.tv_service_status)
                if (tvStatus != null && status != null) {
                    tvStatus.text = "Status: $status"
                }

                // API Dashboard Data (from LyricService inspection)
                if (intent.hasExtra("hasPromotable")) {
                    val hasPromotable = intent.getBooleanExtra("hasPromotable", false)
                    tvNotifCapability.text = "Notif.hasPromotable: $hasPromotable"
                    tvNotifCapability.setTextColor(
                        ContextCompat.getColor(
                            context,
                            if (hasPromotable) R.color.status_active else R.color.status_inactive
                        )
                    )
                }

                if (intent.hasExtra("isPromoted")) {
                    val isPromoted = intent.getBooleanExtra("isPromoted", false)
                    tvNotifFlag.text = "Flag PROMOTED_ONGOING: $isPromoted"
                    tvNotifFlag.setTextColor(
                        ContextCompat.getColor(
                            context,
                            if (isPromoted) R.color.status_active else R.color.status_inactive
                        )
                    )
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Handle Edge-to-Edge / Status Bar Insets
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initializeDefaultWhitelist()
        bindViews()
        setupClickListeners()
        setupObservers()

        checkPromotedNotificationPermission()
        updateVersionInfo()
    }

    // API 36 Permission Check (Standard Runtime Permission)
    private fun checkPromotedNotificationPermission() {
        if (Build.VERSION.SDK_INT >= 36) {
            if (checkSelfPermission("android.permission.POST_PROMOTED_NOTIFICATIONS") != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(arrayOf("android.permission.POST_PROMOTED_NOTIFICATIONS"), 102)
            }
        }
    }

    private fun initializeDefaultWhitelist() {
        val prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
        if (!prefs.contains(PREF_WHITELIST)) {
            val set = HashSet(Arrays.asList(*DEFAULT_WHITELIST))
            prefs.edit().putStringSet(PREF_WHITELIST, set).apply()
        }
    }

    private fun bindViews() {
        cardStatus = findViewById(R.id.cv_status)
        ivStatusIcon = findViewById(R.id.iv_status_icon)
        tvStatusText = findViewById(R.id.tv_status_text)

        tvAppVersion = findViewById(R.id.tv_app_version)

        tvSong = findViewById(R.id.tv_song)
        tvArtist = findViewById(R.id.tv_artist)
        tvLyric = findViewById(R.id.tv_lyric)

        // Dashboard Bindings
        tvApiPermission = findViewById(R.id.tv_api_permission)
        tvNotifCapability = findViewById(R.id.tv_notif_capability)
        tvNotifFlag = findViewById(R.id.tv_notif_flag)
        btnOpenPromotedSettings = findViewById(R.id.btn_open_settings)
        pbProgress = findViewById(R.id.pb_progress)
    }

    private fun setupClickListeners() {
        val cvSettings = findViewById<View>(R.id.cv_settings)
        cvSettings?.setOnClickListener { startActivity(Intent(this, SettingsActivity::class.java)) }

        // Open Promoted Settings
        btnOpenPromotedSettings.setOnClickListener {
            try {
                // Try specific API 36 Settings Action
                val intent = Intent("android.settings.MANAGE_APP_PROMOTED_NOTIFICATIONS")
                intent.putExtra(Settings.EXTRA_APP_PACKAGE, packageName)
                startActivity(intent)
            } catch (e: Exception) {
                // Fallback to App Notification Settings
                val intent = Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS)
                intent.putExtra(Settings.EXTRA_APP_PACKAGE, packageName)
                startActivity(intent)
                Toast.makeText(this, "Promoted Settings not found, opening Notification Settings", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun copyToClipboard(label: String, text: String?) {
        if (!text.isNullOrEmpty() && text != "-") {
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText(label, text)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(this, "$label copied!", Toast.LENGTH_SHORT).show()
        }
    }

    private fun updateStatusCardState() {
        // Check Permission
        val listenerGranted = isNotificationListenerEnabled()

        if (!listenerGranted) {
            // State: Missing Permission
            setCardState(false, "Permission Required", R.color.status_inactive) // Red/Orange
        } else {
            // Check Playing Status
            val isPlaying = LyricRepository.getInstance().isPlaying.value
            if (java.lang.Boolean.TRUE == isPlaying) {
                val source = LyricRepository.getInstance().sourceAppName.value
                setCardState(true, "Active: " + (source ?: "Music"), R.color.status_active)
            } else {
                setCardState(true, "Service Ready (Idle)", R.color.status_active) // Or distinct color?
            }
        }
    }

    private fun setCardState(isActive: Boolean, text: String, colorResId: Int) {
        cardStatus.setCardBackgroundColor(ContextCompat.getColor(this, colorResId))
        tvStatusText.text = text

        if (isActive) {
            ivStatusIcon.setImageResource(R.drawable.ic_check_circle)
        } else {
            ivStatusIcon.setImageResource(R.drawable.ic_cancel)
        }
    }

    private fun isNotificationListenerEnabled(): Boolean {
        return Settings.Secure.getString(contentResolver, "enabled_notification_listeners").contains(packageName)
    }

    private fun updateVersionInfo() {
        try {
            val version = packageManager.getPackageInfo(packageName, 0).versionName
            val type = if ((applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0) "Debug" else "Release"
            tvAppVersion.text = getString(R.string.app_version_fmt, version, type)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    // Check API Status (Reflection)
    private fun checkApiStatusForDashboard() {
        // Only show in Debug builds
        val cvApiStatus = findViewById<View>(R.id.cv_api_status)
        if (!BuildConfig.DEBUG) {
            cvApiStatus.visibility = View.GONE
            return
        }
        cvApiStatus.visibility = View.VISIBLE

        if (Build.VERSION.SDK_INT < 36) {
            tvApiPermission.text = "Permission: N/A (Pre-API 36)"
            tvNotifCapability.text = "Notif.hasPromotable: N/A"
            tvNotifFlag.text = "Flag PROMOTED: N/A"
            return
        }

        try {
            val nm = getSystemService(android.app.NotificationManager::class.java)
            val m = nm.javaClass.getMethod("canPostPromotedNotifications")
            val granted = m.invoke(nm) as Boolean

            if (granted) {
                tvApiPermission.text = "Permission (canPost): GRANTED ✅"
                tvApiPermission.setTextColor(ContextCompat.getColor(this, R.color.status_active))
            } else {
                tvApiPermission.text = "Permission (canPost): DENIED ❌"
                tvApiPermission.setTextColor(ContextCompat.getColor(this, R.color.status_inactive))
            }
        } catch (e: Exception) {
            tvApiPermission.text = "Permission Check Failed: ${e.message}"
        }
    }

    private fun updateDebugProgress() {
        // No-op for now as UI removed
    }

    private fun formatTime(ms: Long): String {
        var seconds = ms / 1000
        val minutes = seconds / 60
        seconds = seconds % 60
        return String.format("%02d:%02d", minutes, seconds)
    }

    private fun setupObservers() {
        val repo = LyricRepository.getInstance()

        // Playback Status Observer
        repo.isPlaying.observe(this) { _ ->
            updateStatusCardState()
        }

        // Lyric Observer
        repo.currentLyric.observe(this) { lyric ->
            if (lyric != null) {
                tvLyric.text = lyric
            } else {
                tvLyric.text = "-"
            }
        }

        // Metadata Observer
        repo.liveMetadata.observe(this) { mediaInfo ->
            if (mediaInfo == null) return@observe
            // Ensure defaults if null
            tvSong.text = mediaInfo.title
            tvArtist.text = mediaInfo.artist
        }
        
        // Progress Observer
        repo.liveProgress.observe(this) { progress ->
            if (progress != null && progress.duration > 0) {
                pbProgress.max = progress.duration.toInt()
                pbProgress.setProgress(progress.position.toInt(), true)
            } else {
                pbProgress.progress = 0
            }
        }
    }

    override fun onResume() {
        super.onResume()
        updateStatusCardState()
        checkApiStatusForDashboard()

        checkApiStatusForDashboard()

        // Register Diag Receiver
        val filter = IntentFilter()
        filter.addAction("com.example.islandlyrics.DIAG_UPDATE")
        filter.addAction("com.example.islandlyrics.STATUS_UPDATE")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(diagReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
        } else {
            registerReceiver(diagReceiver, filter)
        }
    }

    override fun onPause() {
        super.onPause()
        try {
            unregisterReceiver(diagReceiver)
        } catch (e: Exception) {
        }
    }

    companion object {
        private const val TAG = "IslandLyrics"
        private const val PREFS_NAME = "IslandLyricsPrefs"
        private const val PREF_WHITELIST = "whitelist_packages"

        // Default Whitelist
        private val DEFAULT_WHITELIST = arrayOf(
            "com.netease.cloudmusic",
            "com.tencent.qqmusic",
            "com.kugou.android",
            "com.spotify.music",
            "com.apple.android.music",
            "com.google.android.youtube",
            "com.google.android.apps.youtube.music"
        )
    }
}
